### PR TITLE
[L0] Fix kernel error handling and enumeration checking

### PR DESCRIPTION
**Pull request by nrspruit**
- Fixed kernel create to free memory and close with nullptr
- Fixed argument index checking for kernels and argument size checks
- UR_KERNEL_INFO_NUM_REGS to be reported same as UR_KERNEL_INFO_NUM_ARGS

### CODE REVIEW

**Review by pbalcer — APPROVED** [review]
kernel.cpp changes lgtm, but the command buffer changes look suspicious - is that a bad rebase or intentional?

File: source/adapters/level_zero/command_buffer.cpp
Comment (EwanC) [inline source/adapters/level_zero/command_buffer.cpp:1]: These changes look like they are from https://github.com/oneapi-src/unified-runtime/pull/1775 and don't think should be included in this PR.

Comment (nrspruit) [reply source/adapters/level_zero/command_buffer.cpp:1]: the rebase must have broken, I will fix, thanks!

**Review by EwanC — APPROVED** [review]
rebase to remove command-buffer changes LGTM

**Review by winstonzhang-intel — APPROVED** [review]

**Review by pbalcer — APPROVED** [review]

**Pull request merged**